### PR TITLE
Support CLI

### PR DESCRIPTION
**Pull request by wesleytodd**
A command line interface for operating on support metadata.  This is a work in progress PR, and will probably be the last to merge once the js api is finalized.

**TODO:**

- [ ] Setup a support declaration
- [ ] Validate the support schema
- [ ] Output badges markdown from support metadata
- [ ] Show support information
  - [ ] List specific fields from schema
  - [ ] List all support url's from install tree
  - [ ] List information for remote packages

### CODE REVIEW

**Comment by Eomm** [comment]
Did you know oclif?
I used it and it is pretty fast to create a cli setup 

I was wondering to open a pr with that

**Comment by wesleytodd** [comment]
Hey @Eomm, yes I know about oclif.  I have philosophical differences 😛 from the authors of that package and would prefer not to use it.  I just got into work so probably don't have time to write up why I don't like that approach, but I can later if you are interested.

**Comment by Eomm** [comment]
> would prefer not to use it.

No problem, I wanted to point out a possible approach

> if you are interested.

Yes, sure! I tried and except some concern about some customization, I found it easy to setup and test

**Comment by wesleytodd** [comment]
> Yes, sure! I tried and except some concern about some customization, I found it easy to setup and test

That is actually one of my main concerns.  The approach they take is really opposite my preferred approach to libraries which will be critical dependencies.  Strong opinions in libraries can be great when they really fit your project and will not get in your way.  Same with feature rich libraries.  In the case of this package, I am confident that neither the features or opinions will help, and might get in the way.  And we have a GREAT and much more fitting tool for the job, `yargs`!

**Comment by Eomm** [comment]
@wesleytodd can I work on this branch?

**Comment by wesleytodd** [comment]
Absolutely!  Should follow the same guideline for reviews before merging like we do on the node repos?

**Comment by Eomm** [comment]
> Should follow the same guideline for reviews before merging like we do on the node repos?

Not sure it fits a code repo:
+ could the 4 approvals limit too high limit for this org because it is smaller than the package-maintenance group?
+ the 7 days waiting period before merge conflict with a bug or other stuff/chore

In doubt, I would ask the group on the next meeting

**Comment by wesleytodd** [comment]
Yeah I agree that is probably too high, for some reason I thought it was 2 and 48 hours.  So far myself and @dominykas have been the only ones to start new repos here, and for that we didn't do an process.  I personally like that because it means we can move fast.  I don't want to put roadblocks for anyone, but for existing things we need to make sure we get agreement, for example I have already integrated this into other projects, so I want to ensure I can continue to use it as I was (until we all agree to break it in a major version).

It seems like one review, no objections and 48h is good for additions and changes to published packages.  For new ones I would vote for no process but some review once we see what direction it is going if it continues to belong here.  I agree we should bring this up in the next meeting if we have time.

File: lib/cli/commands/validate.js
Comment (ljharb) [inline lib/cli/commands/validate.js]: this doesn’t validate that it’s a path within the package - ie, it doesn’t prevent ..-ing out

File: lib/cli/index.js
Comment (mhdawson) [inline lib/cli/index.js]:
```suggestion
const Loggerr = require('loggerr')
```

Comment (mhdawson) [reply lib/cli/index.js]: On linux the capital does not match the module name which has the lower case `l`

**Review by mhdawson — APPROVED** [review]
LGTM

File: lib/cli/commands/validate.js
Comment (wesleytodd) [reply lib/cli/commands/validate.js]: Should be fixed.

File: lib/cli/index.js
Comment (wesleytodd) [reply lib/cli/index.js]: Fixed!

**Comment by wesleytodd** [comment]
Squashed it down to one conventional commit.  I hope we can strive for clean PRs like this, so wanted to land that before someone else merged it.  FWIW @Eomm I typically will just `git commit --amend` and force push on WIP commits.  If you have more changes you want to land on this, feel free to do the same!

**Comment by wesleytodd** [comment]
Oh, and I should also mention, if you do that you end up with a nice co-author credit on github.  Because I squashed it and was also the original committer it removed you.  Even just rebasing again would be enough to gain back co-author credit, so feel free to do that if it matters to you!

**Comment by ljharb** [comment]
You can add `Co-authored-by: name <email>` lines to the bottom of the commit message to keep the credit - there are zero reasons not to rebase and force push frequently :-D

**Comment by Eomm** [comment]
Don't worry for that @wesleytodd 👍
As said here https://github.com/pkgjs/support/pull/7#issuecomment-628217360 I think we could go ahead 💪

**Comment by wesleytodd** [comment]
@Eomm oops, hadn't seen that comment yet!  Sorry for repeating the topic.

File: lib/cli/commands/validate.js
Comment (mhdawson) [reply lib/cli/commands/validate.js]: I had more completely addressed that in my PR (And now the code have been moved as well so there will likely be conflict)

**Comment by wesleytodd** [comment]
> You can add Co-authored-by: name <email> lines to the bottom of the commit message to keep the credit

WOW! I cant believe I didnt know this. Thanks so much @ljharb!

**Comment by mhdawson** [comment]
@wesleytodd wondering if you think this is ready to land, so we can then look at landing follow on PRs?

File: lib/cli/commands/validate.js
Comment (ljharb) [inline lib/cli/commands/validate.js:15]:
```suggestion
      return yargs;
```
but we can address this later ;-)

File: lib/cli/commands/validate.js
Comment (ljharb) [inline lib/cli/commands/validate.js]: this won't work if it's invoked in a nested dir, but also, find for now

File: lib/cli/commands/validate.js
Comment (ljharb) [inline lib/cli/commands/validate.js]: how does this handle if the string resolves to a symlink to a path outside the cwd?

File: lib/cli/index.js
Comment (ljharb) [inline lib/cli/index.js]:
```suggestion
      describe: 'Supress all output',
```

File: package.json
Comment (ljharb) [inline package.json]: run `npx ls-engines --save` so the "engines" requirements are explicit :-D

**Review by ljharb — APPROVED** [review]
LGTM pending comments

File: lib/cli/commands/validate.js
Comment (Eomm) [reply lib/cli/commands/validate.js:15]: `standard` doesn't validate the `;` at the end of line

File: lib/cli/commands/validate.js
Comment (Eomm) [reply lib/cli/commands/validate.js]: Do you think [pkg-up](npmjs.com/package/pkg-up) would be better for this task?

File: lib/cli/commands/validate.js
Comment (Eomm) [reply lib/cli/commands/validate.js]: I changed how this file is read: now it is relative to the package.json we just read.

Right now the read of the package.json is forced to cwd (so nothing has changed), but if we search the package.json this code will adapt to that path

File: test/cli.js
Comment (Eomm) [inline test/cli.js:13]: I'm not a mocha expert, so if you know a better assertion lib feel free to suggest it 👍

**Comment by mhdawson** [comment]
I'm going to land this and then rebase https://github.com/pkgjs/support/pull/7 on it as the next step from our discussion when we last go together.

File: lib/cli/index.js
Comment (ljharb) [inline lib/cli/index.js]:
```suggestion
      describe: 'Suppress all output',
```

File: package.json
Comment (ljharb) [inline package.json]:
```suggestion
    "prepublishOnly": "npm test",
```
no point in abbreviating something nobody has to type :-)

**Pull request merged**
